hail/src/main/scala/is/hail/sparkextras/RepartitionedOrderedRDD2.scala

```diff
@@ -55,20 +55,74 @@ class RepartitionedOrderedRDD2 private (@transient val prev: RVD, @transient val
     val ordPartition = partition.asInstanceOf[RepartitionedOrderedRDD2Partition]
     val pord = kOrd.intervalEndpointOrdering
     val range = ordPartition.range
-    val ur = new UnsafeRow(typ.rowType)
-    val key = new SelectFieldsRow(ur, typ.kFieldIdx)
-
-    Iterator.single { (ctx: RVDContext) =>
-      ordPartition.parents.iterator
-        .flatMap { parentPartition =>
-          prevCRDD.iterator(parentPartition, context).flatMap(_(ctx))
-        }.dropWhile { ptr =>
-          ur.set(ctx.r, ptr)
-          pord.lt(key, range.left)
-        }.takeWhile { ptr =>
-          ur.set(ctx.r, ptr)
-          pord.lteq(key, range.right)
+
+    Iterator.single { (outerCtx: RVDContext) =>
+      new Iterator[Long] {
+        private[this] val innerCtx = outerCtx.freshContext()
+        private[this] val outerRegion = outerCtx.region
+        private[this] val innerRegion = innerCtx.region
+        private[this] val parentIterator = ordPartition.parents.iterator.flatMap(p => prevCRDD.iterator(p, context).next().apply(innerCtx))
+        private[this] var pulled: Boolean = false
+        private[this] var current: Long =  _
+        private[this] val ur = new UnsafeRow(typ.rowType)
+        private[this] val key = new SelectFieldsRow(ur, typ.kFieldIdx)
+
+        // drop left elements at iterator allocation to avoid extra control flow in hasNext()
+        dropLeft()
+
+        private[this] def dropLeft(): Unit = {
+          var eltLessThan = true
+          while (parentIterator.hasNext && eltLessThan) {
+            innerRegion.clear()
```

Comment:
This looks fishy. You're clearing the producer region after calling `hasNext` but before calling `next`. The `parentIterator` may have produced the element when you called `hasNext`, then you clear the region, then `next` returns the dangling pointer.

One way to fix it:
```scala
@tailrec def dropLeft(): Unit = {
  if (parentIterator.hasNext) {
    pull()
    if (pord.lt(key, range.left) {
      innerRegion.clear()
      dropLeft()
    } else if (pord.gt(key, range.right) {
      end()
    }
  } else {
    end()
  }
}
```

Reply:
ahhhhh, yeah, this has to be it. Thanks Patrick!

Reply:
one of the failing tests is passing, I'll push it up! 🎉

Reply:
Thanks for catching that Patrick.